c/shared/source/uforead/uforead.c

```diff
@@ -1819,9 +1818,153 @@ static void reallocFDArray(ufoCtx h){
+    }
+    if ( cur == 0 ) {
+        xmlFreeDoc(doc);
+//        return ( NULL );
+    }
+    if ((!xmlStrEqual((cur)->name, (const xmlChar *) "dict"))) {
+        fprintf(stderr, "Error reading outermost <dict> in %s.\n", "file");
+        xmlFreeDoc(doc);
+//        return(NULL);
+    }
+
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        keyID = NULL;
+        if ((xmlStrEqual(cur->name, (const xmlChar *) "key"))) {
+            parseKeyName(&keyID, cur);
+            if (keyID == NULL){
+                message(h, "Warning: Encountered empty <key></key>.");
+                cur = cur->next;
+                continue;
+            }
+            keyName = (char*) keyID;
+            cur = cur->next;
+            setFontDictKey(h, keyName, cur);
+        }
+        cur = cur->next;
+    }
+}
+
+static void parseKeyName(unsigned char* *keyID, xmlNodePtr cur){
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        if (!xmlStrEqual(cur->name, (const xmlChar *) "text"))
+            continue;
+        *keyID = xmlNodeGetContent(cur);
+        cur = cur->next;
+    }
+}
+
+static void parseXMLArray(ufoCtx h, xmlNodePtr cur){
+    dnaSET_CNT(h->valueArray, 0);
+    cur = cur->xmlChildrenNode;
+    while (cur != NULL) {
+        void* valueString = parseKeyValue(h, cur);
+        if (valueString != NULL)
+            *dnaNEXT(h->valueArray) = valueString;
+        cur = cur->next;
+    }
+}
+
+static void parseXMLDict(ufoCtx h, xmlNodePtr cur){
+    // go through an XML dict
+    unsigned char* keyID;
+    cur = cur->xmlChildrenNode;
+
+    if (parsingFDArray){
+        currentiFD = currentiFD + 1;
+        h->top.FDArray.cnt = h->top.FDArray.cnt + 1;
+        if (h->top.FDArray.cnt > FDArrayInitSize){ /* Memory needs reallocation*/
+            reallocFDArray(h);
+        }
+        abfFontDict* fd = h->top.FDArray.array + currentiFD;
+        abfInitFontDict(fd);
+    }
+
+    while (cur != NULL) {
+        parseKeyName(&keyID, cur);
+//        cur = cur->next;
+//        void *ptr = parseKeyValue(h, cur);
+
+        char* keyName = (char*) keyID;
+        cur = cur->next;
+        setFontDictKey(h, keyName, cur);
+        cur = cur->next;
+    }
+}
+
+static bool isSimpleKey(xmlNodePtr cur){
+    if (xmlStrEqual(cur->name, (const xmlChar *) "string" ) ||
+        xmlStrEqual(cur->name, (const xmlChar *) "integer") ||
+        xmlStrEqual(cur->name, (const xmlChar *) "real")    ||
+        xmlStrEqual(cur->name, (const xmlChar *) "date")) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+static char* parseKeyError(ufoCtx h, xmlNodePtr cur){
+    if (xmlStrEqual(cur->name, (const xmlChar *) "key"))
+        fatal(h, ufoErrParse, "Encountered empty <key>");  // needs fix
+    else if (xmlStrEqual(cur->name, (const xmlChar *) "array"))
+        fatal(h, ufoErrParse, "Encountered empty <array>");
+    else
+        return NULL;
+}
+
+static void *parseKeyValue(ufoCtx h, xmlNodePtr cur){
```

Comment:
Even though it can be convenient it's best to avoid `void *` unless and until absolutely necessary (that is, don't do it if you don't have to, and if you do need a void pointer don't cast something to it until as late as possible). 

Seems like this function is mostly intended to return a string. Are there other cases?

Reply:
Ah I see, that's good to know. You're right, it is fine to return either an unsigned char* or keep it as an XmlChar* (which is an unsigned char*). I would then have to cast this return value from unsigned to signed char* when assigning the dictionary values in setFontDictKey. DO you think there will be risks with that kind of conversion?

Reply:
All new uses of void* should now have been replaced by char*. 

I believe this can be resolved unless you have additional thoughts